src/common/exec_utils.cc

```diff
@@ -0,0 +1,526 @@
+                                   std::vector<NDArray> *post_temp_dst,
+                                   std::unordered_map<uint32_t, uint32_t> *in_temp_idx_map,
+                                   const std::vector<uint32_t> &mutate_idx) {
+  // populate input blobs
+  SetupDefaultBlobsIn(ndinputs, in_bufs, input_blobs, pre_temp_src, pre_temp_dst,
+                      in_temp_idx_map);
+  // populate output blobs
+  SetupDefaultBlobsOut(ndoutputs, out_bufs, req, output_blobs, post_temp_dst,
+                       post_temp_src);
+  // add mutable inputs to post temp list
+  for (const auto idx : mutate_idx) {
+    auto map_iter = in_temp_idx_map->find(idx);
+    if (map_iter != in_temp_idx_map->end()) {
+      post_temp_src->push_back(pre_temp_dst->at(map_iter->second));
+      post_temp_dst->push_back(ndinputs[idx]);
+    }
+  }
+}
+
+
+void CastNonDefaultStorage(const std::vector<NDArray>& src,
+                                  const std::vector<NDArray>& dst,
+                                  const OpContext& ctx,
+                                  const bool is_gpu) {
+  CHECK_EQ(dst.size(), src.size());
+  for (size_t i = 0; i < src.size(); i++) {
+    if (is_gpu) {
+#if MXNET_USE_CUDA
+      CastStorageDispatch<gpu>(ctx, src[i], dst[i]);
+#else
+      LOG(FATAL) << MXNET_GPU_NOT_ENABLED_ERROR;
+#endif
+    } else {
+      CastStorageDispatch<cpu>(ctx, src[i], dst[i]);
+    }
+  }
+}
+
+
+bool SameType(const nnvm::NodeAttrs& attrs,
+                     std::vector<int> *iattr,
+                     std::vector<int> *oattr) {
+  int def_v = -1;
+  for (int v : *oattr) {
+    if (v != -1) {
+      def_v = v; break;
+    }
+  }
+  if (def_v == -1) {
+    for (int v : *iattr) {
+      if (v != -1) {
+        def_v = v; break;
+      }
+    }
+  }
+  if (def_v == -1) return false;
+  for (int& v : *oattr) {
+    v = def_v;
+  }
+  for (int& v : *iattr) {
+    v = def_v;
+  }
+  return true;
+}
+
+
+
+bool DefaultStorageType(const nnvm::NodeAttrs& attrs,
+                               const int dev_mask,
+                               DispatchMode* dispatch_mode,
+                               std::vector<int> *iattr,
+                               std::vector<int> *oattr) {
+  bool fallback = false;
+  for (int& v : *oattr) {
+    if (v == -1) v = kDefaultStorage;
+    if (v != kDefaultStorage) fallback = true;
+  }
+  for (int& v : *iattr) {
+    if (v == -1) v = kDefaultStorage;
+    if (v != kDefaultStorage) fallback = true;
+  }
+  if (*dispatch_mode == DispatchMode::kUndefined) {
+    if (fallback) {
+      *dispatch_mode = DispatchMode::kFComputeFallback;
+    } else {
+      *dispatch_mode = DispatchMode::kFCompute;
+    }
+  }
+  return true;
+}
+
+std::string storage_str(int storage_id) {
+  std::string str;
+  if (storage_id == -1) {
+    str = "var (-1)";
+  } else if (storage_id == -2) {
+    str = "external storage (-2)";
+  } else {
+    str = "group " + std::to_string(storage_id);
+  }
+  return str;
+}
+
+
```

Comment:
nit: remove extra blank line

Reply:
Please address @haojin2 comments before resolving it?